File: Sources/Shared/Protocol.swift
import Foundation

// MARK: - Request

public struct Request: Codable {
    public let id: String
    public let command: String
    public let args: CommandArgs
    public let options: RequestOptions?
    
    public init(id: String, command: String, args: CommandArgs, options: RequestOptions? = nil) {
        self.id = id
        self.command = command
        self.args = args
        self.options = options
    }
}

public struct RequestOptions: Codable {
    public var timeout: Int?
    public var json: Bool?
    public var verbose: Bool?
    
    public init(timeout: Int? = nil, json: Bool? = nil, verbose: Bool? = nil) {
        self.timeout = timeout
        self.json = json
        self.verbose = verbose
    }
}

// MARK: - Command Args

public enum CommandArgs: Codable {
    case snapshot(SnapshotArgs)
    case click(ClickArgs)
    case fill(FillArgs)
    case type(TypeArgs)
    case press(PressArgs)
    case scroll(ScrollArgs)
    case screenshot(ScreenshotArgs)
    case open(OpenArgs)
    case get(GetArgs)
    case status
    
    private enum CodingKeys: String, CodingKey {
        case type
    }
    
    // Custom encoding to flatten for JSON protocol
    public func encode(to encoder: Encoder) throws {
        switch self {
        case .snapshot(let args): try args.encode(to: encoder)
        case .click(let args): try args.encode(to: encoder)
        case .fill(let args): try args.encode(to: encoder)
        case .type(let args): try args.encode(to: encoder)
        case .press(let args): try args.encode(to: encoder)
        case .scroll(let args): try args.encode(to: encoder)
        case .screenshot(let args): try args.encode(to: encoder)
        case .open(let args): try args.encode(to: encoder)
        case .get(let args): try args.encode(to: encoder)
        case .status: try [String: String]().encode(to: encoder)
        }
    }
    
    public init(from decoder: Decoder) throws {
        // Decoding is done by the daemon based on command name, not auto-decoded
        fatalError("CommandArgs should be decoded manually based on command name")
    }
}

// MARK: - Specific Arg Types

public struct SnapshotArgs: Codable {
    public var interactive: Bool
    public var compact: Bool
    public var depth: Int?
    public var app: String?
    
    public init(interactive: Bool = true, compact: Bool = false, depth: Int? = nil, app: String? = nil) {
        self.interactive = interactive
        self.compact = compact
        self.depth = depth
        self.app = app
    }
}

public struct ClickArgs: Codable {
    public var ref: String?
    public var x: Double?
    public var y: Double?
    public var double: Bool
    public var right: Bool
    
    public init(ref: String? = nil, x: Double? = nil, y: Double? = nil, double: Bool = false, right: Bool = false) {
        self.ref = ref
        self.x = x
        self.y = y
        self.double = double
        self.right = right
    }
}

public struct FillArgs: Codable {
    public var ref: String
    public var text: String
    
    public init(ref: String, text: String) {
        self.ref = ref
        self.text = text
    }
}

public struct TypeArgs: Codable {
    public var ref: String?
    public var text: String
    
    public init(ref: String? = nil, text: String) {
        self.ref = ref
        self.text = text
    }
}

public struct PressArgs: Codable {
    public var key: String
    public var modifiers: [String]?
    
    public init(key: String, modifiers: [String]? = nil) {
        self.key = key
        self.modifiers = modifiers
    }
}

public struct ScrollArgs: Codable {
    public var direction: String
    public var amount: Int?
    public var ref: String?
    
    public init(direction: String, amount: Int? = nil, ref: String? = nil) {
        self.direction = direction
        self.amount = amount
        self.ref = ref
    }
}

public struct ScreenshotArgs: Codable {
    public var full: Bool
    public var app: String?
    
    public init(full: Bool = false, app: String? = nil) {
        self.full = full
        self.app = app
    }
}

public struct OpenArgs: Codable {
    public var target: String
    
    public init(target: String) {
        self.target = target
    }
}

public struct GetArgs: Codable {
    public var what: String  // "text", "title", "apps", "windows"
    public var ref: String?
    public var app: String?
    
    public init(what: String, ref: String? = nil, app: String? = nil) {
        self.what = what
        self.ref = ref
        self.app = app
    }
}

// MARK: - Response

public struct Response: Codable {
    public let id: String
    public let success: Bool
    public let data: ResponseData?
    public let error: ErrorInfo?
    public let timing: Timing?
    
    public init(id: String, success: Bool, data: ResponseData? = nil, error: ErrorInfo? = nil, timing: Timing? = nil) {
        self.id = id
        self.success = success
        self.data = data
        self.error = error
        self.timing = timing
    }
    
    public static func ok(id: String, data: ResponseData, elapsed: Double) -> Response {
        Response(id: id, success: true, data: data, timing: Timing(elapsed_ms: elapsed))
    }
    
    public static func fail(id: String, error: ErrorInfo, elapsed: Double = 0) -> Response {
        Response(id: id, success: false, error: error, timing: Timing(elapsed_ms: elapsed))
    }
}

public struct Timing: Codable {
    public let elapsed_ms: Double
    
    public init(elapsed_ms: Double) {
        self.elapsed_ms = elapsed_ms
    }
}

// MARK: - Response Data

public enum ResponseData: Codable {
    case snapshot(SnapshotData)
    case click(ClickData)
    case fill(FillData)
    case type(TypeData)
    case press(PressData)
    case scroll(ScrollData)
    case screenshot(ScreenshotData)
    case open(OpenData)
    case getApps(GetAppsData)
    case getText(GetTextData)
    case status(StatusData)
    case raw(String) // fallback
    
    private enum DiscriminatorKey: String, CodingKey { case _type }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: DiscriminatorKey.self)
        switch self {
        case .snapshot(let d): try container.encode("snapshot", forKey: ._type); try d.encode(to: encoder)
        case .click(let d): try container.encode("click", forKey: ._type); try d.encode(to: encoder)
        case .fill(let d): try container.encode("fill", forKey: ._type); try d.encode(to: encoder)
        case .type(let d): try container.encode("type", forKey: ._type); try d.encode(to: encoder)
        case .press(let d): try container.encode("press", forKey: ._type); try d.encode(to: encoder)
        case .scroll(let d): try container.encode("scroll", forKey: ._type); try d.encode(to: encoder)
        case .screenshot(let d): try container.encode("screenshot", forKey: ._type); try d.encode(to: encoder)
        case .open(let d): try container.encode("open", forKey: ._type); try d.encode(to: encoder)
        case .getApps(let d): try container.encode("getApps", forKey: ._type); try d.encode(to: encoder)
        case .getText(let d): try container.encode("getText", forKey: ._type); try d.encode(to: encoder)
        case .status(let d): try container.encode("status", forKey: ._type); try d.encode(to: encoder)
        case .raw(let s): try container.encode("raw", forKey: ._type); try s.encode(to: encoder)
        }
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DiscriminatorKey.self)
        let type = try container.decode(String.self, forKey: ._type)
        switch type {
        case "snapshot": self = .snapshot(try SnapshotData(from: decoder))
        case "click": self = .click(try ClickData(from: decoder))
        case "fill": self = .fill(try FillData(from: decoder))
        case "type": self = .type(try TypeData(from: decoder))
        case "press": self = .press(try PressData(from: decoder))
        case "scroll": self = .scroll(try ScrollData(from: decoder))
        case "screenshot": self = .screenshot(try ScreenshotData(from: decoder))
        case "open": self = .open(try OpenData(from: decoder))
        case "getApps": self = .getApps(try GetAppsData(from: decoder))
        case "getText": self = .getText(try GetTextData(from: decoder))
        case "status": self = .status(try StatusData(from: decoder))
        default: self = .raw(type)
        }
    }
}

// MARK: - Data Payloads

public struct SnapshotData: Codable {
    public var text: String
    public var refCount: Int
    public var app: String
    public var window: String?
    
    public init(text: String, refCount: Int, app: String, window: String? = nil) {
        self.text = text
        self.refCount = refCount
        self.app = app
        self.window = window
    }
}

public struct ClickData: Codable {
    public var ref: String?
    public var coordinates: Point
    public var element: ElementInfo?
    
    public init(ref: String? = nil, coordinates: Point, element: ElementInfo? = nil) {
        self.ref = ref
        self.coordinates = coordinates
        self.element = element
    }
}

public struct FillData: Codable {
    public var ref: String
    public var text: String
    
    public init(ref: String, text: String) {
        self.ref = ref
        self.text = text
    }
}

public struct TypeData: Codable {
    public var ref: String?
    public var text: String
    
    public init(ref: String? = nil, text: String) {
        self.ref = ref
        self.text = text
    }
}

public struct PressData: Codable {
    public var key: String
    public var modifiers: [String]
    
    public init(key: String, modifiers: [String] = []) {
        self.key = key
        self.modifiers = modifiers
    }
}

public struct ScrollData: Codable {
    public var direction: String
    public var amount: Int
    
    public init(direction: String, amount: Int) {
        self.direction = direction
        self.amount = amount
    }
}

public struct ScreenshotData: Codable {
    public var path: String
    public var width: Int
    public var height: Int
    public var scale: Int
    
    public init(path: String, width: Int, height: Int, scale: Int = 1) {
        self.path = path
        self.width = width
        self.height = height
        self.scale = scale
    }
}

public struct OpenData: Codable {
    public var app: String
    public var pid: Int
    public var wasRunning: Bool
    
    public init(app: String, pid: Int, wasRunning: Bool) {
        self.app = app
        self.pid = pid
        self.wasRunning = wasRunning
    }
}

public struct GetAppsData: Codable {
    public var apps: [AppInfo]
    
    public init(apps: [AppInfo]) {
        self.apps = apps
    }
}

public struct GetTextData: Codable {
    public var ref: String?
    public var text: String
    
    public init(ref: String? = nil, text: String) {
        self.ref = ref
        self.text = text
    }
}

public struct StatusData: Codable {
    public var daemonPid: Int
    public var accessibilityPermission: Bool
    public var screenRecordingPermission: Bool
    public var frontmostApp: String?
    public var frontmostPid: Int?
    public var frontmostWindow: String?
    public var refMapCount: Int
    public var refMapAgeMs: Double?
    
    public init(daemonPid: Int, accessibilityPermission: Bool, screenRecordingPermission: Bool,
                frontmostApp: String? = nil, frontmostPid: Int? = nil, frontmostWindow: String? = nil,
                refMapCount: Int = 0, refMapAgeMs: Double? = nil) {
        self.daemonPid = daemonPid
        self.accessibilityPermission = accessibilityPermission
        self.screenRecordingPermission = screenRecordingPermission
        self.frontmostApp = frontmostApp
        self.frontmostPid = frontmostPid
        self.frontmostWindow = frontmostWindow
        self.refMapCount = refMapCount
        self.refMapAgeMs = refMapAgeMs
    }
}

// MARK: - Supporting Types

public struct Point: Codable {
    public var x: Double
    public var y: Double
    
    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public struct ElementInfo: Codable {
    public var role: String
    public var label: String?
    
    public init(role: String, label: String? = nil) {
        self.role = role
        self.label = label
    }
}

public struct AppInfo: Codable {
    public var name: String
    public var pid: Int
    public var isActive: Bool
    
    public init(name: String, pid: Int, isActive: Bool) {
        self.name = name
        self.pid = pid
        self.isActive = isActive
    }
}

File: crates/shared/src/protocol.rs
use serde::{Deserialize, Serialize};

// MARK: - Request

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Request {
    pub id: String,
    pub command: String,
    pub args: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<RequestOptions>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub json: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verbose: Option<bool>,
}

// MARK: - Command Args

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnapshotArgs {
    #[serde(default = "default_true")]
    pub interactive: bool,
    #[serde(default)]
    pub compact: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub depth: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selector: Option<String>,
}

fn default_true() -> bool { true }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClickArgs {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub x: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub y: Option<f64>,
    #[serde(default)]
    pub double: bool,
    #[serde(default)]
    pub right: bool,
    /// Bring app to foreground (required for coordinate clicks in --app mode)
    #[serde(default)]
    pub foreground: bool,
    /// Target app name (for headless click routing)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
    /// Skip post-click wait (for SPA navigation)
    #[serde(default)]
    pub no_wait: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FillArgs {
    pub r#ref: String,
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TypeArgs {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PressArgs {
    pub key: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifiers: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScrollArgs {
    pub direction: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub amount: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenshotArgs {
    #[serde(default)]
    pub full: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenArgs {
    pub target: String,
    #[serde(default)]
    pub with_cdp: bool,
    #[serde(default)]
    pub background: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetArgs {
    pub what: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WaitArgs {
    /// Element @ref (e.g. @e5) or milliseconds (e.g. "2000")
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ref_or_ms: Option<String>,
    /// Wait for page load state: "networkidle", "domcontentloaded", "load"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub load: Option<String>,
    /// Target app name
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
}

// MARK: - Response

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Response {
    pub id: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<ResponseData>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<crate::types::ErrorInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timing: Option<Timing>,
}

impl Response {
    pub fn ok(id: String, data: ResponseData, elapsed_ms: f64) -> Self {
        Self { id, success: true, data: Some(data), error: None, timing: Some(Timing { elapsed_ms }) }
    }
    pub fn fail(id: String, error: crate::types::ErrorInfo, elapsed_ms: f64) -> Self {
        Self { id, success: false, data: None, error: Some(error), timing: Some(Timing { elapsed_ms }) }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Timing {
    pub elapsed_ms: f64,
}

// MARK: - Response Data

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "_type")]
pub enum ResponseData {
    #[serde(rename = "snapshot")]
    Snapshot(SnapshotData),
    #[serde(rename = "click")]
    Click(ClickData),
    #[serde(rename = "fill")]
    Fill(FillData),
    #[serde(rename = "type")]
    Type(TypeData),
    #[serde(rename = "press")]
    Press(PressData),
    #[serde(rename = "scroll")]
    Scroll(ScrollData),
    #[serde(rename = "screenshot")]
    Screenshot(ScreenshotData),
    #[serde(rename = "open")]
    Open(OpenData),
    #[serde(rename = "getApps")]
    GetApps(GetAppsData),
    #[serde(rename = "getText")]
    GetText(GetTextData),
    #[serde(rename = "status")]
    Status(StatusData),
    #[serde(rename = "wait")]
    Wait(WaitData),
}

// MARK: - Data Payloads

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnapshotData {
    pub text: String,
    pub ref_count: i32,
    pub app: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub window: Option<String>,
    /// Profiling report (only present when --verbose is used)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub profile: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClickData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    pub coordinates: Point,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub element: Option<ElementInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FillData {
    pub r#ref: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TypeData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PressData {
    pub key: String,
    #[serde(default)]
    pub modifiers: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScrollData {
    pub direction: String,
    pub amount: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenshotData {
    pub path: String,
    pub width: u32,
    pub height: u32,
    #[serde(default = "default_one")]
    pub scale: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub window_origin_x: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub window_origin_y: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app_name: Option<String>,
}

fn default_one() -> u32 { 1 }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenData {
    pub app: String,
    pub pid: i32,
    pub was_running: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cdp_port: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetAppsData {
    pub apps: Vec<AppInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetTextData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WaitData {
    pub waited_ms: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StatusData {
    pub daemon_pid: i32,
    pub accessibility_permission: bool,
    pub screen_recording_permission: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frontmost_app: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frontmost_pid: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frontmost_window: Option<String>,
    pub ref_map_count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ref_map_age_ms: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_cdp_connections: Option<i32>,
}

// MARK: - Supporting Types

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Point {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ElementInfo {
    pub role: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppInfo {
    pub name: String,
    pub pid: i32,
    pub is_active: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_request_json_roundtrip() {
        let req = Request {
            id: "test_1".to_string(),
            command: "snapshot".to_string(),
            args: serde_json::json!({"interactive": true, "depth": 10}),
            options: None,
        };
        let json = serde_json::to_string(&req).unwrap();
        let decoded: Request = serde_json::from_str(&json).unwrap();
        assert_eq!(decoded.id, "test_1");
        assert_eq!(decoded.command, "snapshot");
    }

    #[test]
    fn test_response_ok_roundtrip() {
        let resp = Response::ok(
            "r1".to_string(),
            ResponseData::Status(StatusData {
                daemon_pid: 1234,
                accessibility_permission: true,
                screen_recording_permission: true,
                frontmost_app: Some("Finder".to_string()),
                frontmost_pid: Some(456),
                frontmost_window: None,
                ref_map_count: 0,
                ref_map_age_ms: None,
                active_cdp_connections: Some(0),
            }),
            15.5,
        );
        let json = serde_json::to_string(&resp).unwrap();
        let decoded: Response = serde_json::from_str(&json).unwrap();
        assert!(decoded.success);
        assert_eq!(decoded.id, "r1");
    }

    #[test]
    fn test_response_fail_roundtrip() {
        let resp = Response::fail(
            "r2".to_string(),
            crate::types::ErrorInfo {
                code: "REF_NOT_FOUND".to_string(),
                message: "Element @e3 not found.".to_string(),
                suggestion: Some("Run `snapshot` to refresh.".to_string()),
            },
            2.1,
        );
        let json = serde_json::to_string(&resp).unwrap();
        let decoded: Response = serde_json::from_str(&json).unwrap();
        assert!(!decoded.success);
        assert!(decoded.error.is_some());
    }
}
